File: DynamicProgramming/Introduction_to_dynamic_problems/Fibonacci_Number_with_Map.java
import java.util.HashMap;
import java.util.Map;

class Fibonacci {
	// Function to find nth Fibonacci number
	public static int fib(int n, Map<Integer, Integer> lookup) {
		if (n <= 1) {
			return n;
		}

		// If sub-problem is seen for the first time
		if (!lookup.containsKey(n)) {
			int val = fib(n - 1, lookup) + fib(n - 2, lookup);
			lookup.put(n, val);
		}

		return lookup.get(n);
	}

	public static void main(String[] args) {
		int n = 8;
		Map<Integer, Integer> lookup = new HashMap<>();

		System.out.println(fib(n, lookup));
	}
}

File: DynamicProgramming/Introduction_to_dynamic_problems/Fibonacci_Number_with_Map.py
# Function to find nth Fibonacci number
def fib(n, lookup):

    if n <= 1:
        return n

    # If sub-problem is seen for the first time
    if n not in lookup:
        val = fib(n - 1, lookup) + fib(n - 2, lookup)
        lookup[n] = val

    return lookup[n]


if __name__ == "__main__":

    n = 8
    lookup = {}

    print(fib(n, lookup))
